compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs

```diff
@@ -829,14 +831,25 @@ checkTemplateKey param tcon TemplateKey{..} = do
       checkExpr tplKeyBody tplKeyType
     checkExpr tplKeyMaintainers (tplKeyType :-> TList TParty)
 
+checkDefException :: MonadGamma m => Module -> DefException -> m ()
+checkDefException m DefException{..} = do
+    let modName = moduleName m
+        tcon = Qualified PRSelf modName exnName
+    DefDataType _loc _name _serializable tyParams dataCons <- inWorld (lookupDataType tcon)
+    unless (null tyParams) $ throwWithContext (EExpectedExceptionTypeHasNoParams modName exnName)
+    _ <- match _DataRecord (EExpectedExceptionTypeIsRecord modName exnName) dataCons
+    case NM.lookup exnName (moduleTemplates m) of
+        Nothing -> pure ()
+        Just _ -> throwWithContext (EExpectedExceptionTypeIsNotTemplate modName exnName)
+
 -- NOTE(MH): It is important that the data type definitions are checked first.
 -- The type checker for expressions relies on the fact that data type
 -- definitions do _not_ contain free variables.
 checkModule :: MonadGamma m => Module -> m ()
-checkModule m@(Module _modName _path _flags synonyms dataTypes values templates _exceptions) = do
+checkModule m@(Module _modName _path _flags synonyms dataTypes values templates exceptions) = do
   let with ctx f x = withContext (ctx x) (f x)
   traverse_ (with (ContextDefTypeSyn m) checkDefTypeSyn) synonyms
   traverse_ (with (ContextDefDataType m) checkDefDataType) dataTypes
   traverse_ (with (\t -> ContextTemplate m t TPWhole) $ checkTemplate m) templates
   traverse_ (with (ContextDefValue m) checkDefValue) values
-  -- TODO #8020: check exception types are well formed and don't overlap with templates
+  traverse_ (with (ContextDefException m) (checkDefException m)) exceptions
```

Comment:
It's probably better to move this before checking any expressions such that we've run through this before we call `checkExceptionType`. It's not strictly necessary but I think it makes for better error messages.

Reply:
Oh, good catch!